File: heap-sort/python/heap_sort.py
import random


def sort(data):
    build_max_heap(data)

    print(data)

    for i in reversed(range(1, len(data))):
        data[0], data[i] = data[i], data[0]

        heapify(data, 0, i - 1)

        print(data)


def build_max_heap(data):
    for i in reversed(range(len(data) // 2)):
        heapify(data, i, len(data) - 1)


def heapify(data, target_idx, heap_idx):
    largest_idx = target_idx

    left_idx = 2 * target_idx + 1
    if left_idx <= heap_idx and data[left_idx] > data[largest_idx]:
        largest_idx = left_idx

    right_idx = 2 * target_idx + 2
    if right_idx <= heap_idx and data[right_idx] > data[largest_idx]:
        largest_idx = right_idx

    if largest_idx != target_idx:
        data[target_idx], data[largest_idx] = data[largest_idx], data[target_idx]

        heapify(data, largest_idx, heap_idx)


if __name__ == "__main__":
    data = [random.randint(1, 20) for _ in range(20)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: heap-sort/golang/heap_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func sort(data []int) {
	buildMaxHeap(data)

	fmt.Println(data)

	for i := len(data) - 1; i > 0; i-- {
		data[0], data[i] = data[i], data[0]

		heapify(data, 0, i-1)

		fmt.Println(data)
	}
}

func buildMaxHeap(data []int) {
	for i := len(data) / 2; i >= 0; i-- {
		heapify(data, i, len(data)-1)
	}
}

func heapify(data []int, targetIdx, heapIdx int) {
	largestIdx := targetIdx

	leftIdx := 2*targetIdx + 1
	if leftIdx <= heapIdx && data[leftIdx] > data[largestIdx] {
		largestIdx = leftIdx
	}

	rightIdx := 2 * (targetIdx + 1)
	if rightIdx <= heapIdx && data[rightIdx] > data[largestIdx] {
		largestIdx = rightIdx
	}

	if largestIdx != targetIdx {
		data[targetIdx], data[largestIdx] = data[largestIdx], data[targetIdx]

		heapify(data, largestIdx, heapIdx)
	}
}

func main() {
	data := make([]int, 20)

	for i := range 20 {
		data[i] = rand.Intn(20) + 1
	}

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
